### PR TITLE
Task/ABSA

### PR DESCRIPTION
**Pull request by Takerdat23**
Add aspect based sentiment analysis task code for RNN and transformers

### CODE REVIEW

File: models/RNN_ABSA.py [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+                dropout=self.dropout_prob if self.num_layer > 1 else 0,
+            )
+        
+
+        # Dropout layer
+        self.dropout = nn.Dropout(self.dropout_prob)
+
+        # ABSA output head 
+
+        self.outputHead = Aspect_Based_SA_Output(config.dropout  , self.d_model * self.bidirectional, config.num_output, config.num_categories )
```
Comment (phusroyal) [inline models/RNN_ABSA.py]: shoud not use directly from config

File: models/RNN_ViPher_ABSA.py [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+                dropout=self.dropout_prob if self.num_layer > 1 else 0,
+            )
+            
+        self.num_labels = config.num_output
+
+        # Dropout layer
+        self.dropout = nn.Dropout(self.dropout_prob)
+
+        # ABSA output head 
+        self.outputHead = Aspect_Based_SA_Output(config.dropout  , self.d_model * self.bidirectional , config.num_output, config.num_categories )
```
Comment (phusroyal) [inline models/RNN_ViPher_ABSA.py]: shoud not use directly from config

File: tasks/aspect_based_classification.py [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+
+                logits, _ = self.model(input_ids, label)
+                output = logits.argmax(dim=-1).long()
+                
+                output = output.cpu().numpy()
+                label = label.cpu().numpy()
+                
+                for pred , true in zip(output , label): 
+          
+                    predictions_Aspect = [1 if sentiment != 0 else 0 for idx , sentiment in enumerate(pred) ]
```
Comment (phusroyal) [inline tasks/aspect_based_classification.py]: 1. `1 if sentiment != 0 else 0` This is true when there is only two labels in aspect.
2. Why use enumerate when you don't use idx?
3. Naming conventions. This is aspect why use sentiment?

File: tasks/aspect_based_classification.py [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+                    ground_truths_Aspect = [1 if sentiment != 0 else 0 for idx ,  sentiment in enumerate(true) ]
```
Comment (phusroyal) [inline tasks/aspect_based_classification.py]: Same as above

File: tasks/aspect_based_classification.py [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+
+                    predictions_Sentiment = [idx for idx in pred]
```
Comment (phusroyal) [inline tasks/aspect_based_classification.py]: Ugly code! list(pred) or pred.tolist()

File: tasks/aspect_based_classification.py [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+        aspect_score = {metric: float(value) for metric, value in aspect_score.items()}
+
+        # Overall sentiment classification scores
+        sentiment_score = self.compute_scores(all_sentiment_pred, all_sentiment_label)
+        sentiment_score = {metric: float(value) for metric, value in sentiment_score.items()}
+
+        return {
+            'aspect': aspect_score,
+            'sentiment': sentiment_score
+        }
```
Comment (phusroyal) [inline tasks/aspect_based_classification.py:161]: should create two dictionaries then store separately so the score_collectors can also work with

File: vocabs/BPE.py [hunk 7]
```diff
@@ -87,6 +90,10 @@ def make_vocab(self, config):
                     for label in item[config.label]:
                         label = label.split("-")[-1]
                         labels.add(label)
+                elif self.config.get("task_type", None) == "aspect_based":
+                    for label in item["label"]: 
+                        aspects.add(label['aspect'])
+                        sentiments.add(label['sentiment'])
```
Comment (phusroyal) [inline vocabs/BPE.py]: these two lines of 95,96 are okay, for now! may need future maintenance if the data store differently

File: vocabs/BPE.py [hunk 8]
```diff
@@ -101,12 +108,27 @@ def make_vocab(self, config):
...
-        labels = sorted(list(labels))
-        self.i2l = {i: label for i, label in enumerate(labels)}
-        self.l2i = {label: i for i, label in enumerate(labels)}
-
+        
+        if self.config.get("task_type", None) == "aspect_based":
+            aspects = list(aspects)
+            sentiments = list(sentiments)
+            
+            aspects = list(aspects)
```
Comment (phusroyal) [inline vocabs/BPE.py]: duplicate and need sort to have a consistent ordering

File: vocabs/BPE.py [hunk 9]
```diff
@@ -101,12 +108,27 @@ def make_vocab(self, config):
...
+            self.i2a = {i: label for i, label in enumerate(aspects)}
+            self.a2i = {label: i for i, label in enumerate(aspects)}
```
Comment (phusroyal) [inline vocabs/BPE.py:119]: 117,118: naming conventions

File: vocabs/BPE.py [hunk 10]
```diff
@@ -101,12 +108,27 @@ def make_vocab(self, config):
...
+            
+            sentiments = list(sentiments)
+            self.i2s = {i: label for i, label in enumerate(sentiments, 1)}
+            self.i2s[0] = None
+            self.s2i = {label: i for i, label in enumerate(sentiments, 1)}
```
Comment (phusroyal) [inline vocabs/BPE.py:124]: 121,123: naming conventions

File: vocabs/BPE.py [hunk 11]
```diff
@@ -101,12 +108,27 @@ def make_vocab(self, config):
...
+        
+        if self.config.get("task_type", None) == "aspect_based":
+            aspects = list(aspects)
+            sentiments = list(sentiments)
+            
+            aspects = list(aspects)
+            self.i2a = {i: label for i, label in enumerate(aspects)}
+            self.a2i = {label: i for i, label in enumerate(aspects)}
+            
+            sentiments = list(sentiments)
```
Comment (phusroyal) [inline vocabs/BPE.py]: duplicate and need sort to have a consistent ordering

File: vocabs/BPE.py [hunk 12]
```diff
@@ -306,6 +328,18 @@ def encode_label(self, label: str) -> torch.Tensor:
...
  
             return torch.Tensor(labels).long()
+        elif self.config.get("task_type", None) == "aspect_based":
+            
+            label_vector = torch.zeros(self.total_aspects_labels["aspects"])
+            for l in label: 
+                aspect = l['aspect']
+                sentiment = l['sentiment']
+                # active the OTHERS case
+                if aspect == "OTHERS":
```
Comment (phusroyal) [inline vocabs/BPE.py:340]: in visfd when the aspect == other the sentiment always equals to null, so why don't we drop them in the first place

**Review by phusroyal — CHANGES_REQUESTED** [review]

File: tasks/aspect_based_classification.py (hunk 3 above)
Comment (Takerdat23) [reply tasks/aspect_based_classification.py]: the pred size is [B, Num_categories], so the 1 in position [idx] means aspect number [idx] exist and 0 when the aspect number [idx] does not exist.

File: tasks/aspect_based_classification.py [hunk 13]
```diff
@@ -174,9 +174,9 @@ def get_predictions(self, dataset):
         predictions = []
         results = []
         test_scores = self.evaluate_metrics(self.test_dataloader)
-        # val_scores = self.evaluate_metrics(self.dev_dataloader)
+        val_scores = self.evaluate_metrics(self.test_dataloader)
         scores.append({
-            # "val_scores": val_scores , 
+            "val_scores": val_scores ,
```
Comment (phusroyal) [inline tasks/aspect_based_classification.py]: To maintain consistency with other experiments, we don't record val_scores.

**Review by phusroyal — CHANGES_REQUESTED** [review]
